src/Microsoft.IdentityModel.Tokens/SecurityTokenDescriptor.cs

```diff
@@ -12,10 +16,61 @@ namespace Microsoft.IdentityModel.Tokens
     /// </summary>
     public class SecurityTokenDescriptor
     {
+        // TODO: At next major version, remove _audienceSet, _audiencesSet, UseAudiences, Audience, and relevant logic.
+        private bool _audienceSet = false;
+        private bool _audiencesSet = false;
+        private IList<string> _audiences;
+        internal bool UseAudiences => _audiencesSet;
+        internal bool HasAudience => _audiences != null && _audiences.Count > 0;
```

Comment:
Can we merge Audience and Audiences into one if both are set when creating tokens?

Reply:
I thought about doing that but it seemed like it would introduce a lot of complexity around expected behavior and possibly concurrency. 

A couple of examples around expected behavior: if I set Audience then later set Audiences, or vice versa should the audience(s) be added to the existing audience(s) or replace it/them, we could have a set and just keep adding, but what if Audience is set to A, Audiences is added and has A, B, C later I set Audience to D, do I leave A in or take it out? If I set Audiences to null do I expect that Audience is still filled out or that all Audiences are set to null?

With any concurrency it further complicates things as in order to guarantee behavior we'd need to ensure atomic interactions between the two properties.

@GeoK do you have a design in mind that would alleviate these sorts of concerns? Requiring one or the other but not both was my best shot at it.

Reply:
SecurityTokenDescriptor is used exclusively for token creation. I'd imagine that we could do something similar to what is done on the validation side - if we see ValidIssuer and ValidIssuers we would concat and consider both.

If one sets both Audience and Audiences, we'd add all distinct values as audiences on a token. That would eliminate the need for taking care of properties that are set, particularly if Audience will get deprecated.

Reply:
What was the resolution on this one? I see it marked as resolved, but no comment on the outcome or decision.

Reply:
In a meeting Brent told us not to make any changes to the values in the SecurityTokenDescriptor as they are owned by the customer and should remain as they are set.